lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1048,6 +1051,127 @@ bool PreCheckExpression::walkToClosureExprPre(ClosureExpr *closure) {
+
+        // The first component is the parent type.
+        auto *ParentComp = new (TC.Context) SimpleIdentTypeRepr(
+          DRE->getNameLoc().getBaseNameLoc(), TD->getName());
+        ParentComp->setValue(TD);
+        Components.push_back(ParentComp);
+
+        // The second component is the member we just found.
+        auto *NewComp = new (TC.Context) SimpleIdentTypeRepr(NameLoc, Name);
+        NewComp->setValue(Result.front().first);
+        Components.push_back(NewComp);
+
+        auto *NewTypeRepr = IdentTypeRepr::create(TC.Context, Components);
+        return new (TC.Context) TypeExpr(TypeLoc(NewTypeRepr, Type()));
+      }
+    }
+
+    return nullptr;
+  }
+
+  auto *TyExpr = dyn_cast<TypeExpr>(UDE->getBase());
+  if (!TyExpr)
+    return nullptr;
+
+  auto *InnerTypeRepr = TyExpr->getTypeRepr();
+  if (!InnerTypeRepr)
+    return nullptr;
+
+  // Fold 'T.Protocol' into a protocol metatype.
+  if (Name == TC.Context.Id_Protocol) {
+    auto *NewTypeRepr =
+      new (TC.Context) ProtocolTypeRepr(InnerTypeRepr, NameLoc);
+    return new (TC.Context) TypeExpr(TypeLoc(NewTypeRepr, Type()));
+  }
+
+  // Fold 'T.Type' into an existential metatype if 'T' is a protocol,
+  // or an ordinary metatype otherwise.
+  if (Name == TC.Context.Id_Type) {
+    auto *NewTypeRepr =
+      new (TC.Context) MetatypeTypeRepr(InnerTypeRepr, NameLoc);
+    return new (TC.Context) TypeExpr(TypeLoc(NewTypeRepr, Type()));
+  }
+
+  // Fold 'T.U' into a nested type.
+  //
+  // FIXME: Also support this if 'T' is a generic parameter.
+  if (auto *ITR = dyn_cast<IdentTypeRepr>(InnerTypeRepr)) {
+    // The last component should be bound already.
+    if (auto *D = ITR->getComponentRange().back()->getBoundDecl()) {
+      auto BaseTy = D->getDeclaredInterfaceType();
+      // The last component should be a module, nominal type, or a typealias
+      // whose underlying type is a nominal type, otherwise we cannot perform
+      // a type member lookup, which will be diagnosed later.
+      if (BaseTy->getAnyNominal() ||
+          BaseTy->is<ModuleType>()) {
+        // See if there is a member type with this name.
+        auto Result = TC.lookupMemberType(DC,
+                                          BaseTy,
+                                          Name,
+                                          defaultMemberLookupOptions);
+
+        // If there is no nested type with this name, we have a lookup of
+        // a non-type member, so leave the expression as-is.
+        if (Result.size() == 1) {
+          auto *NewDecl = Result.front().first;
+          // Filter out dicey cases that we diagnose or handle later:
+          //
+          // 1) Dependent typealias or associated type reference with
+          //    protocol base.
+          if (NewDecl->getDeclaredInterfaceType()->hasTypeParameter() &&
+              BaseTy->isExistentialType())
+            return nullptr;
+
+          // 2) Generic typealias reference with unbound generic base.
```

Comment:
I'm going to refactor things so that the "filtering" isn't necessary. You're right that it's fragile.